dwave/system/temperatures.py

```diff
@@ -0,0 +1,488 @@
+# Copyright 2021 D-Wave Systems Inc.
+#
+#    Licensed under the Apache License, Version 2.0 (the "License");
+#    you may not use this file except in compliance with the License.
+#    You may obtain a copy of the License at
+#
+#        http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+
+"""Effective temperature estimators
+
+Maximum pseudo-likelihood is an efficient estimator for the temperature 
+describing a classical Boltzmann distribution P(x) = exp(-H(x)/T)/Z(T) 
+given samples from that distribution, where H(x) is the classical energy 
+function. This estimator is implemented.
+See also: https://doi.org/10.3389/fict.2016.00023 
+https://www.jstor.org/stable/25464568 # doctest: +SKIP
+
+A temperature can also be inferred from an assumed freeze-out phenomena in 
+combination with the schedule energy scale [B(s), the energy of the problem 
+Hamiltonian]. This estimator is implemented. Relevant device-specific properties
+are published for online solvers:
+https://docs.dwavesys.com/docs/latest/doc_physical_properties.html
+
+""" 
+
+import warnings
+import numpy as np
+import dimod
+from scipy import optimize
+
+__all__ = ['effective_field', 'maximum_pseudolikelihood_temperature',
+           'freezeout_effective_temperature', 'fast_effective_temperature']
+
+def effective_field(bqm,
+                    samples_like=None,
+                    current_state_energy=False):
+    '''Returns the effective field for all variables and all samples.
+    
+    The effective field with current_state_energy = False is the energy
+    attributable to setting a variable to value 1, conditioned on fixed values 
+    for all neighboring variables (relative to exclusion of the variable, and 
+    associated energy terms, from the problem). 
+    
+    The effective field with current_state_energy = True is the energy gained
+    by flipping the variable state against its current value (from say -1 to 1 
+    in the Ising case, or 0 to 1 in the QUBO case). A positive value indicates 
+    that the energy can be decreased by flipping the variable, hence the 
+    variable is in a locally excited state.
+    If all values are negative  (positive) within a sample, that sample is a 
+    local minima (maxima).  
```

Comment:
This only considers energy changes possible with respect to single bit flips, the energy change conditioned on the other N-1 variables being fixed. It might be that the energy can be lowered by considering a correlated flip of many spins. E.g. H = -(sum_i s_i)^2 - eps*sum_I s_i. A ferromagnet in a small (eps) external field. All -1 is a local minima, all +1 is a global minima. In the first case all the local fields are of value -(N-1) + eps (negative for small eps), in the latter case they are of value -(N-1) - eps (also negative at small eps).